File: csharp/820.short-encoding-of-words.cs
/*
 * @lc app=leetcode id=820 lang=csharp
 *
 * [820] Short Encoding of Words
 */

using System.Collections.Generic;
using System.Linq;

// @lc code=start
public class Solution {
    public int MinimumLengthEncoding(string[] words) {
        var set = new HashSet<string>(words);
        var remove = new HashSet<string>(
            set.SelectMany(word => Enumerable.Range(1, word.Length - 1).Select(i => word.Substring(i))).Where(set.Contains)
        );
        return set.Where(word => !remove.Contains(word)).Select(word => word.Length + 1).Sum();
    }
}
// @lc code=end

File: java/820.short-encoding-of-words.java
/*
 * @lc app=leetcode id=820 lang=java
 *
 * [820] Short Encoding of Words
 */

// @lc code=start
class Solution {
    public int minimumLengthEncoding(String[] words) {
        Set<String> set = new HashSet<String>(Arrays.asList(words)), remove = new HashSet<String>();
        for (String word : set) {
            if (remove.contains(word)) {
                continue;
            }
            int k = word.length();
            for (int i = 1; i < k; ++i) {
                String s = word.substring(k - i);
                if (set.contains(s)) {
                    remove.add(s);
                }
            }
        }
        int ret = 0;
        for (String word : set) {
            if (!remove.contains(word)) {
                ret += word.length() + 1;
            }
        }
        return ret;
    }
}
// @lc code=end

